src/FlightController.cpp

```diff
@@ -4,106 +4,256 @@
 
 namespace fcu {
 
-FlightController::FlightController(const std::string &device, const size_t baudrate) {
-    client.connect(device, baudrate);
-    client.start();
+FlightController::FlightController() : 
+    msp_version_(1), armed_(false), control_source_(ControlSource::NONE),
+    msp_timer_(std::bind(&FlightController::generateMSP,this),0.1)
+{
+    
 }
 
 FlightController::~FlightController() {
-    client.stop();
-}
-
-void FlightController::waitForConnection() {
-    std::cout<<"Wait for FC..."<<std::endl;
-    msp::msg::Ident ident;
-    while(!client.request(ident, 0.5));
-    std::cout<<"MultiWii version "<< size_t(ident.version)<<" ready"<<std::endl;
+    disconnect();
 }
 
-void FlightController::initialise() {
-    // wait for connection to be established
-    while(client.request(ident, 0.5)==-1);
 
-    msp::msg::ApiVersion api;
-    if(client.request(api)) {
-        // this is Cleanflight
-        firmware = FirmwareType::CLEANFLIGHT;
-        std::cout<<"Cleanflight API "<<api.major<<"."<<api.minor<<"."<<api.protocol<<" ready"<<std::endl;
+bool FlightController::connect(const std::string &device, const size_t baudrate, const double& timeout) {
+    if (!client_.start(device,baudrate)) return false;
+    
+    int rc;
+    msp::msg::FcVariant fcvar(fw_variant_);
+    rc = client_.sendMessage(fcvar,timeout);
+    if ( rc == 1) {
+        fw_variant_ = msp::variant_map[fcvar.identifier()];
+        std::cout << fcvar;
     }
-    else {
-        // this is MultiWii
-        firmware = FirmwareType::MULTIWII;
-        std::cout<<"MultiWii version "<< size_t(ident.version)<<" ready"<<std::endl;
+    else std::cout << rc << std::endl;
+    
+    if (fw_variant_ != msp::FirmwareVariant::MWII) {
+        msp::msg::ApiVersion api_version(fw_variant_);
+        if(client_.sendMessage(api_version,timeout)) {
+            std::cout << api_version;
+            msp_version_ = api_version.major();
+            client_.setVersion(msp_version_);
+        }
     }
-
-    // get sensors
-    msp::msg::Status status;
-    client.request(status);
-    sensors = status.sensors;
-
+    
+    msp::msg::FcVersion fcver(fw_variant_);
+    rc = client_.sendMessage(fcver,timeout);
+    if (rc == 1) {
+        std::cout << fcver;
+    }
+    else std::cout << rc << std::endl;
```

Comment:
We don't need to print the return value.

Reply:
done